src/NServiceBus.Transport.Msmq/MessagePump.cs

```diff
@@ -78,49 +83,54 @@ public Task StartReceive(CancellationToken cancellationToken)
         {
             MessageQueue.ClearConnectionCache();
 
-            cancellationTokenSource = new CancellationTokenSource();
-            this.cancellationToken = cancellationTokenSource.Token;
-
             // LongRunning is useless combined with async/await
-            messagePumpTask = Task.Run(() => ProcessMessages(), CancellationToken.None);
+            messagePumpTask = Task.Run(() => PumpMessages(), cancellationToken);
 
             return Task.CompletedTask;
         }
 
         public async Task StopReceive(CancellationToken cancellationToken)
         {
-            cancellationTokenSource.Cancel();
+            messagePumpCancellationTokenSource?.Cancel();
+            cancellationToken.Register(() => messageProcessingCancellationTokenSource?.Cancel());
```

Comment:
could this throw an `ObjectDisposedException` if the triggers after line 123? Seems like this could be avoided by setting the tokensource value to `null` after disposing it?